src/internal/core.rs

```diff
@@ -190,16 +189,46 @@ impl<DP: DependencyProvider> State<DP> {
         // If there are no more changed packages, unit propagation is done.
-        Ok(root_causes)
+        Ok(satisfier_causes)
     }
 
     /// Return the root cause or the terminal incompatibility.
     /// CF <https://github.com/dart-lang/pub/blob/master/doc/solver.md#unit-propagation>
+    ///
+    /// Usually by the time we have a conflict `unit_propagation` has done a lot of work.
+    /// So the actual conflict we find is important, but not particularly actionable.
+    /// It says something like "the dependency on package X and the dependency on package Y are incompatible".
+    /// To make it actionable we want to track it back to decisions that made the dependency required.
+    /// "The decision on B is incompatible with the decision on C,
+    ///  because unit propagation from just those decisions will lead to the conflict about X and Y"
+    /// is much more actionable, backtrack until one of those decisions can be revisited.
+    /// To make a practical, we really only need one of the terms to be a decision.
+    /// We may as well leave the other terms general. Something like
+    /// "the dependency on the package X is incompatible with the decision on C" tends to work out pretty well.
+    /// Then if A turns out to also have a dependency on X the resulting root cause is still useful.
+    /// Of course, this is more heuristics than science. If the output is too general, then `unit_propagation` will
+    /// handle the confusion by calling us again with the next most specific conflict it comes across.
+    /// If the output is to specific, then the outer `solver` loop will eventually end up calling us again
+    /// until all possibilities are enumerated.
+    ///
+    /// This function combines incompatibilities with things that make the problem inevitable to end up with a
+    /// more useful incompatibility. For the correctness of the PubGrub algorithm only the final output is required.
+    /// By banning the final output, unit propagation will prevent the intermediate steps from occurring again,
+    /// at least prevent the exact same way. However, the statistics collected for `prioritize`may want
+    /// to analyze those intermediate steps. For example we might start with "there is no version 1 of Z",
+    /// and `conflict_resolution` may be able to determine that "that was inevitable when we picked version 1 of X"
+    /// which was inevitable when picked W and ... and version 1 of B, which was depended on by version 1 of A.
+    /// Therefore the root cause may simplify all the way down to "we cannot pick version 1 of A".
+    /// This will prevent us going down this path again. However when we start looking at version 2 of A,
+    /// and discover that it depends on version 2 of B, we will want to prioritize the chain of intermediate steps
+    /// to confirm if it has a problem with the same shape.
+    /// The `satisfier_causes` argument keeps track of these intermediate steps so that the caller can use.
```

Comment:
Overall I like the edits. I have a nit about

> Say we found that the decision for X and the decision for Y are incompatible.

I'm going to quibble about the term `decision` in this sentence. In my brain `decision`  implies something that became constrained because of a call to `choose_version`/`add_decision`. Because we check for conflicts in `add_version` simple cases end up handled there. We generally do not end up here with a conflict that is already in terms of decisions. Instead  We start with something abstract like `X (in some range) depends on Y (in some range)`. It's the job of this function to peel back the layers, figure out why the partial solution has an assignment (which is not necessarily decision) that is disjoint with the mentioned range for Y and why the partial solution has an assignment that is a subset of the mentioned range for X.

> How will we apply the root cause?

When we discover that A  has a dependency on X `unit_propagation` will rescan the root cause 
`"the dependency on the package X is incompatible with the decision on C"`, and remove that version of C from the available  range stored in partial solution.

> In which direction are "too general" and "too specific" to be understood here?

I intended for to specific to mean referring to exact decided versions. "B == 123 is incompatible with C == 456" would be very specific. Extremely actionable, `unit_propagation`  needs to do very little work to figure out if it's relevant. But also not likely to  be reusable. We are likely to end up with `O(# versions of B * # versions of C)`  of these incompatibilities laying around. "any version of X is incompatible with any version of Y"  would be very general. There will only ever need to be one, and it will be relevant  and useful any time X or Y  come up again. But, it takes a lot of work to figure out how it relates to decisions or where we need to backtrack to.

Obviously the wording of the paragraphs needs  work, thank you for helping me with it!